fastcrawler/core/spider.py

```diff
@@ -215,28 +281,38 @@ async def start(self, silent: bool = True) -> None:
             await self.start_up()
             await self.async_init()
             current_depth = 0
-            async with self.engine(connection_limit=self.engine_request_limit) as session:
+            async with self.engine(connection_limit=self._engine_request_limit) as session:
                 while await self.control_condition(current_depth):
-                    current_depth += 1
                     urls = list(await self.get_urls())
-                    for idx in range(0, len(urls), self.engine_request_limit):
-                        end_index = idx + self.engine_request_limit
-                        batch_urls = urls[idx:end_index]
-                        requests = [Request(url=url) for url in batch_urls]
-                        responses = await self.requests(session, requests)
-                        self.update_crawl_urls(batch_urls)
-                        for response in responses.values():
-                            self.remove_url_from_pending(response)
-
-                        results = [
-                            self.parse_response(response)
-                            for response in responses.values()
-                            if response
-                        ]
-                        await self.save(results)
+                    for idx in range(0, len(urls), self.get_batch_size):
```

Comment:
too nested

Reply:
indeed, all would goo to Batch class. 
I haven't prototyped it yet. 
Currently vahid would be working on core, with that new feature that we discussed yesterday.
Then i'll kick off implementing the Batch class.